### PR TITLE
[release-18.0] Direct PR. Fix regression where reference tables with a different name on sharded keyspaces were not routed correctly.

### PR DESCRIPTION
**Pull request by rohit-nayak-ps**
## Description

If a reference table is named differently in the sharded keyspace, inserts and updates to that table behave differently than if they had the same name. Looks like there was a regression somewhere between v16 and v18.

The only code change is in the dml planning, where tables are rerouted. Additionally unit tests and an e2e test has been added to validate the rerouting and protect against regression.

Note: since there has been significant refactors in the vtgate planner between v18 and the current main, this is being implemented as a standalone PR instead of following the backport route. https://github.com/vitessio/vitess/pull/15796 will fix the issue in v19 and v20.

Test cases have been added for dmls with reference tables have been added:
* with the same name as the source
* with a different name as the source
* with the different name as the source, but with an alias

## Related Issue(s)

https://github.com/vitessio/vitess/issues/15770

## Checklist

-   [ ] "Backport to:" labels have been added if this change should be back-ported to release branches
-   [ ] If this change is to be back-ported to previous releases, a justification is included in the PR description
-   [ ] Tests were added or are not required
-   [ ] Did the new or modified tests pass consistently locally and on CI?
-   [ ] Documentation was added or is not required

### CODE REVIEW

File: go/test/endtoend/vreplication/reference_test.go [hunk 1]
```diff
@@ -0,0 +1,131 @@
...
+	catchup(t, tab80Dash, "wfCat", "Materialize Category")
+	catchup(t, tabDash80, "wfMfg", "Materialize Manufacturer")
+	catchup(t, tab80Dash, "wfMfg", "Materialize Manufacturer")
+
+	vtgateConn.Close()
+	vtgateConn = getConnection(t, vc.ClusterConfig.hostname, vc.ClusterConfig.vtgateMySQLPort)
+	defer vtgateConn.Close()
+	waitForRowCount(t, vtgateConn, sks, "cat", 3)
+	waitForRowCount(t, vtgateConn, sks, "mfg2", 3)
+
```
Comment (rohit-nayak-ps) [inline go/test/endtoend/vreplication/reference_test.go:119]: The first two queries fail, including the insert into `mfg`. That works if we use the same name for the reference table: in this test we are using a different name `mfg2` for the reference table whose source name is `mfg`.

File: go/test/endtoend/vreplication/helper_test.go [hunk 2]
```diff
@@ -463,7 +463,7 @@ func validateDryRunResults(t *testing.T, output string, want []string) {
 			w = strings.TrimSpace(w[1:])
 			result := strings.HasPrefix(g, w)
 			match = result
-			//t.Logf("Partial match |%v|%v|%v\n", w, g, match)
+			// t.Logf("Partial match |%v|%v|%v\n", w, g, match)
```
Comment (harshit-gangal) [inline go/test/endtoend/vreplication/helper_test.go:466]: nit: this log is a comment, can be removed.

File: go/vt/vtgate/planbuilder/dml_planner.go [hunk 3]
```diff
@@ -42,7 +43,22 @@ func rewriteRoutedTables(stmt sqlparser.Statement, vschema plancontext.VSchema)
...
+			}
+			if vindexTbl != nil {
+				// vindex cannot be present in a dml statement.
+				return false, vterrors.VT09014()
+			}
+			tableName.Qualifier = vschemaTable.GetTableName().Qualifier
+			tableName.Name = sqlparser.NewIdentifierCS(vschemaTable.Name.String())
+			newAliasTbl := sqlparser.NewAliasedTableExpr(tableName, "")
+			aliasTbl.Expr = newAliasTbl.Expr
+			aliasTbl.As = newAliasTbl.As
```
Comment (harshit-gangal) [inline go/vt/vtgate/planbuilder/dml_planner.go:60]: The alias should be either an existing one or the reference table name. Should not be source table name.

Comment (harshit-gangal) [reply go/vt/vtgate/planbuilder/dml_planner.go:60]: better to add some tests around it in `dml_cases.json`

Comment (harshit-gangal) [reply go/vt/vtgate/planbuilder/dml_planner.go:60]: I see you have created a new test file for it. So, those can be added there.
DML with alias.

Comment (rohit-nayak-ps) [reply go/vt/vtgate/planbuilder/dml_planner.go:60]: Moved the test cases into reference_cases.json.

Comment (rohit-nayak-ps) [reply go/vt/vtgate/planbuilder/dml_planner.go:60]: > The alias should be either an existing one or the reference table name. Should not be source table name.

So I should only change the qualifier and retain the existing alias' table name?

**Comment by deepthi** [comment]
@vitessio/query-serving does this need to be back ported to v17?

File: go/vt/vtgate/planbuilder/dml_planner.go (hunk 3 above)
Comment (rohit-nayak-ps) [reply go/vt/vtgate/planbuilder/dml_planner.go]: @harshit-gangal, updated the code as we discussed and added more test cases for queries where aliases are specified.

**Review by GuptaManan100 — APPROVED** [review]

**Review by frouioui — APPROVED** [review]

**Comment by harshit-gangal** [comment]
> @vitessio/query-serving does this need to be back ported to v17?

According to me, this DML routing of the reference table was added in v17. I am not sure about the regression referred to here.

File: go/test/endtoend/vreplication/helper_test.go [hunk 4]
```diff
@@ -463,7 +463,6 @@ func validateDryRunResults(t *testing.T, output string, want []string) {
 			w = strings.TrimSpace(w[1:])
 			result := strings.HasPrefix(g, w)
 			match = result
-			//t.Logf("Partial match |%v|%v|%v\n", w, g, match)
```
Comment (harshit-gangal) [inline go/test/endtoend/vreplication/helper_test.go:466]: nit: should be removed or uncommented.

File: go/vt/vtgate/planbuilder/dml_planner.go [hunk 5]
```diff
@@ -42,10 +43,27 @@ func rewriteRoutedTables(stmt sqlparser.Statement, vschema plancontext.VSchema)
...
+		// For reference tables, managed by Vitess (Source not nil), we need to use the source table as the target.
+		if vschemaTable.Type == vindexes.TypeReference && vschemaTable.Source != nil {
+			vschemaTable, vindexTbl, _, _, _, err = vschema.FindTableOrVindex(vschemaTable.Source.TableName)
+			if err != nil {
+				return false, err
+			}
+			if vindexTbl != nil {
+				// vindex cannot be present in a dml statement.
+				return false, vterrors.VT09014()
+			}
```
Comment (harshit-gangal) [inline go/vt/vtgate/planbuilder/dml_planner.go:55]: not sure if this will happen? do you have a test case for this?

**Review by harshit-gangal — APPROVED** [review]

File: go/vt/vtgate/planbuilder/dml_planner.go (hunk 5 above)
Comment (rohit-nayak-ps) [reply go/vt/vtgate/planbuilder/dml_planner.go:55]: You mean the nil vindexTbl check? The same check is made above in the same function. So I assumed it was required. I can remove it, but is there any problem adding the extra check?

Comment (rohit-nayak-ps) [reply go/vt/vtgate/planbuilder/dml_planner.go:55]: Merging as is for now: assuming an additional check is safe. Can remove later on if required.

File: go/test/endtoend/vreplication/helper_test.go (hunk 4 above)
Comment (rohit-nayak-ps) [reply go/test/endtoend/vreplication/helper_test.go:466]: Will remove in the next vrep. PR.

**Pull request merged**